### PR TITLE
S2583: NullCoalesceOperator ??= for ConditionEvaluatesToConstant

### LINKED ISSUE #3066: S2583: NullCoalesceOperator ??= for ConditionEvaluatesToConstant
### Description

Add support for `??=` operator to rule S2583 and S2589 and update RSPEC .

This point was removed from S3240 specification in favor of these rules:

> * removing the {{??}} when the left operand is known to be {{null}}, or known to be definitely not {{null}}.

It should be applied to SE rules and extended to ??= as well.

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Fixes #3066

Please review RSPEC text update as well.

I've closed the spin-off LVA #3076 because it's not relevant anymore.

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ConditionEvaluatesToConstant.cs [hunk 1]
```diff
@@ -330,5 +350,49 @@ bool IsBooleanLiteral(SyntaxNode syntaxNode)
...
+            }
+            var rightOperands = args.Instruction.DescendantNodesAndSelf()
+                .Select(x => (x is BinaryExpressionSyntax binary && binary.IsKind(SyntaxKind.CoalesceExpression) ? binary.Right : null)
+                          ?? (x is AssignmentExpressionSyntax assign && assign.IsKind(SyntaxKindEx.CoalesceAssignmentExpression) ? assign.Right : null))
+                .WhereNotNull().ToArray();
+            foreach (var right in rightOperands)
+            {
+                //It's too late for left operands of ??= or x=x??, symbolic value was already set
+                ProcessOperand(right, false);
+            }
```
Comment (christophe-zurn-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ConditionEvaluatesToConstant.cs]: Why do we need to look at the descendants here ? Can't we just check the current instruction ? since we will evaluate all instructions, it seems many syntax nodes will be reevaluated many times currently

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/ConditionEvaluatesToConstant.CSharp6.cs [hunk 2]
```diff
@@ -2562,6 +2562,61 @@ public string Method8(string message)
...
+            ret = "Lorem " + (isNull ?? a) + " ipsum"; // Noncompliant
+
+            //Right operand: isNull is known to be null, therefore ?? is useless
+            ret = a ?? null;    // Noncompliant
+            ret = a ?? isNull;  // Noncompliant {{Change this expression which always evaluates to 'null'.}}
+//                     ^^^^^^
+
+            //Combo/Fatality
+            ret = notNull ?? isNull;    //Noncompliant [LeftA, RightA]
+            ret = isNull ?? null;       //Noncompliant [LeftB, RightB]
```
Comment (christophe-zurn-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/ConditionEvaluatesToConstant.CSharp6.cs]: I think these cases are out of scope of the rule. The rule is only about conditions. Although the right hand side is indeed useless, it doesn't really correspond what the rule is supposed to check

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ConditionEvaluatesToConstant.cs [hunk 3]
```diff
@@ -85,18 +93,17 @@ public sealed class ConditionEvaluatesToConstant : SonarDiagnosticAnalyzer
         };
 
         private const string S2583DiagnosticId = "S2583"; // Bug
-        private const string S2583MessageFormat = "Change this condition so that it does not always evaluate to '{0}'; some subsequent code is never executed.";
+        private const string S2583MessageFormatBool = "Change this condition so that it does not always evaluate to '{0}'; some subsequent code is never executed.";
+        private const string S2583MessageFormatNull = "Change this expression which always evaluates to '{0}'.";
```
Comment (christophe-zurn-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ConditionEvaluatesToConstant.cs]: I'd keep the `; some subsequent code is never executed.` part of the message, as the rule S2583 is really about unreachable code

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ConditionEvaluatesToConstant.cs [hunk 4]
```diff
@@ -132,6 +145,16 @@ private static void CheckForRedundantConditions(CSharpExplodedGraph explodedGrap
                         .Except(conditionTrue)
                         .Where(c => !IsInsideCatchOrFinallyBlock(c))
                         .Select(node => GetDiagnostics(node, false)))
+                    .Union(isNull
+                        .Except(isUnknown)
+                        .Except(isNotNull)
+                        .Where(c => !IsInsideCatchOrFinallyBlock(c))
+                        .Select(node => Diagnostic.Create(s2583Null, node.GetLocation(), messageArgs: "null")))
```
Comment (christophe-zurn-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ConditionEvaluatesToConstant.cs]: For this case, I'd raise the `S2589` version of the check, as the right hand side of the code will always be executed, so the condition is useless. This falls in the code smell rule case

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ConditionEvaluatesToConstant.cs [hunk 5]
```diff
@@ -85,18 +93,17 @@ public sealed class ConditionEvaluatesToConstant : SonarDiagnosticAnalyzer
...
         private const string S2589DiagnosticId = "S2589"; // Code smell
         private const string S2589MessageFormat = "Change this condition so that it does not always evaluate to '{0}'.";
 
-        private static readonly DiagnosticDescriptor s2583 =
-            DiagnosticDescriptorBuilder.GetDescriptor(S2583DiagnosticId, S2583MessageFormat, RspecStrings.ResourceManager);
-
-        private static readonly DiagnosticDescriptor s2589 =
-            DiagnosticDescriptorBuilder.GetDescriptor(S2589DiagnosticId, S2589MessageFormat, RspecStrings.ResourceManager);
+        private static readonly DiagnosticDescriptor s2583Bool = DiagnosticDescriptorBuilder.GetDescriptor(S2583DiagnosticId, S2583MessageFormatBool, RspecStrings.ResourceManager);
+        private static readonly DiagnosticDescriptor s2583Null = DiagnosticDescriptorBuilder.GetDescriptor(S2583DiagnosticId, S2583MessageFormatNull, RspecStrings.ResourceManager);
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ConditionEvaluatesToConstant.cs]: We need to fix this one as well `ConsumeValueTaskCorrectly.cs` - that's what I've found as example.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ConditionEvaluatesToConstant.cs (hunk 1 above)
Comment (pavel-mikula-sonarsource) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ConditionEvaluatesToConstant.cs]: That doesn't work for `ret = a ?? null`, because instruction itself is AssignmentExpression here and there's no separate instruction for "a ?? null" in the CFG. I can check AssignmentExpression but I'm afraid that `??` can be subexpression of any kind of expression here.
DescendantNodesAndSelf will produce a lot of reevaluations and I'm hoping for small chunks here.

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/ConditionEvaluatesToConstant.CSharp6.cs (hunk 2 above)
Comment (pavel-mikula-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/ConditionEvaluatesToConstant.CSharp6.cs]: These two cases are only for documentation purposes to see that it will be raised twice on the same line. Both left and right side are useless in this case.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ConditionEvaluatesToConstant.cs (hunk 3 above)
Comment (pavel-mikula-sonarsource) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ConditionEvaluatesToConstant.cs]: The rule title is `Conditionally executed blocks should be reachable`. Shouldn't we update it to `Conditionally executed code should be reachable`? In order to cover expressions with ?? and ??= as well as no-block single liners:
```
if(false)
  DoSomething();
```
Blocks seems to restrictive, not broad enough. But the rename would affect other languages as well - can we do it?

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/ConditionEvaluatesToConstant.CSharp6.cs (hunk 2 above)
Comment (christophe-zurn-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/ConditionEvaluatesToConstant.CSharp6.cs]: Yes. My point was that the fact that the right hand side is useless is not related to this rule, the lines (2593, 2594)
```
ret = a ?? null;
ret = a ?? isNull;
```
should not raise an issue in my POV, as here, the fact that the right hand side is always null is not related to a condition, which seems to me out of scope of the rule.
Same for right hand side of:
```
ret = notNull ?? isNull;
ret = isNull ?? null;
```

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ConditionEvaluatesToConstant.cs (hunk 3 above)
Comment (christophe-zurn-sonarsource) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ConditionEvaluatesToConstant.cs]: Makes sense to me

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/ConditionEvaluatesToConstant.CSharp6.cs (hunk 2 above)
Comment (pavel-mikula-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/ConditionEvaluatesToConstant.CSharp6.cs]: So should I drop the check? Or do we have a better *EvaluatesToConstant rule to use?
I agree that this check does not fulfill the root purpose of the rule. It's collateral additional value only.

File: sonaranalyzer-dotnet/rspec/cs/S2589_c#.html [hunk 6]
```diff
@@ -2,20 +2,27 @@
...
-{
-  DoSomething();
+    if (c || !a) // Noncompliant; "!a" is always "false"
+    {
+        DoSomething();
+    }
+
+    string d = null, res;
+    res = s ?? d; // Noncompliant, d is always null
+    res = d ?? "value"; // Noncompliant, d is always null
```
Comment (pavel-mikula-sonarsource) [inline sonaranalyzer-dotnet/rspec/cs/S2589_c#.html]: Do not merge yet. I need to update Compliant version.

File: sonaranalyzer-dotnet/rspec/cs/S2583_c#.html [hunk 7]
```diff
@@ -2,19 +2,25 @@
...
+    {
+        DoSomething();
+    }
+    else
+    {
+        DoSomethingElse(); // never executed
+    }
+
+    var d = "xxx";
+    var res = d ?? "value"; // Noncompliant, d is always not null, "value" is never used
```
Comment (pavel-mikula-sonarsource) [inline sonaranalyzer-dotnet/rspec/cs/S2583_c#.html]: Do not merge yet. Compliant version was missing in RSPEC.

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/ConditionEvaluatesToConstant.CSharp6.cs (hunk 2 above)
Comment (pavel-mikula-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/ConditionEvaluatesToConstant.CSharp6.cs]: CC @andrei-epure-sonarsource 
Conclusion after discussion with @nicolas-harraudeau-sonarsource is to keep this feature. 
Nicolas pointed out a new POV on this case:
```
ret = a ?? null;
```
the condition `a==null` is **the** useless condition here. The final product will always be the same (value of `a`) no matter value comes in. Therefore it is a "condition" issue and it fits the rule title and description.

I will update RSPEC noncompliant example to make it clear.

File: sonaranalyzer-dotnet/rspec/cs/S2589_c#.html (hunk 6 above)
Comment (pavel-mikula-sonarsource) [reply sonaranalyzer-dotnet/rspec/cs/S2589_c#.html]: Done

File: sonaranalyzer-dotnet/rspec/cs/S2583_c#.html (hunk 7 above)
Comment (pavel-mikula-sonarsource) [reply sonaranalyzer-dotnet/rspec/cs/S2583_c#.html]: Done

**Comment by pavel-mikula-sonarsource** [comment]
Regarding the `?? isNull` situation, I'was trying to make message and marked location clear and consistent. But I left it in the original way:
```
var d = null;
var v2 = s ?? d; {{Change this expression which always evaluates to "null".}}
              ^
```
It would be nice to mark the _condition_, but in every case it's not clear that 'd' is the root cause of this issue.

> Change this condition so that it does not always evaluate to 'FIXME: s? left operand? How to simply explain d?

```
var v2 = s ?? d;
         ^
var v2 = s ?? d;
         ^^^^
var v2 = s ?? d;
         ^^^^^^
```

File: sonaranalyzer-dotnet/rspec/cs/S2589_c#.html [hunk 8]
```diff
@@ -2,38 +2,51 @@
...
-{
-  DoSomething();
+    if (c || !a) // Noncompliant; "!a" is always "false"
+    {
+        DoSomething();
+    }
+
+    string d = null;
+    var v1 = d ?? "value"; // Noncompliant, d is always null
+    var v2 = s ?? d; // Noncompliant, d is always null and produced value is always equal to s. Condition to check value of s is gratuitous.
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/rspec/cs/S2589_c#.html]: should use the definite article "the produced value"

Comment (andrei-epure-sonarsource) [reply sonaranalyzer-dotnet/rspec/cs/S2589_c#.html]: "The condition to check the value of s is gratuitous." or "Checking the value of s is gratuitous."

Comment (pavel-mikula-sonarsource) [reply sonaranalyzer-dotnet/rspec/cs/S2589_c#.html]: RSPEC is updated. Files will be updated as part of release process.

Comment (andrei-epure-sonarsource) [reply sonaranalyzer-dotnet/rspec/cs/S2589_c#.html]: that was quick :)
